custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go

```diff
@@ -0,0 +1,163 @@
+	}
+	project := fmt.Sprintf("projects/%s", t.config.Project)
+	endTime := t.clock.Now()
+	startTime := endTime.Add(-t.reqWindow)
+	filter, err := t.createFilter([]string{fmt.Sprintf("\"%s\"", resourceID)}, t.config.Project, t.config.Zone, t.config.Cluster, metricName)
+	if err != nil {
+		return nil, err
+	}
+	return t.service.Projects.TimeSeries.List(project).Filter(filter).
+		IntervalStartTime(startTime.Format("2006-01-02T15:04:05Z")).
+		IntervalEndTime(endTime.Format("2006-01-02T15:04:05Z")).
+		AggregationPerSeriesAligner("ALIGN_NEXT_OLDER").
+		AggregationAlignmentPeriod(fmt.Sprintf("%vs", int64(t.reqWindow.Seconds()))), nil
+}
+
+// GetRespForNamespacedObject returns translates Stackdriver response to a Custom Metric associated
+// with a single namespaced object.
+func (t *Translator) GetRespForNamespacedObject(response *stackdriver.ListTimeSeriesResponse, groupResource schema.GroupResource, metricName string, namespace string, name string) (*custom_metrics.MetricValue, error) {
+	values, err := t.getMetricValuesFromResponse(groupResource, namespace, response)
+	if err != nil {
+		return nil, err
+	}
+	if len(values) != 1 {
+		return nil, fmt.Errorf("Expected exactly one value for pod %s in namespace %s, but received %v values", name, namespace, len(values))
+	}
+	for _, value := range values {
+		metricValue, err := t.metricFor(value, groupResource, namespace, name, metricName)
+		if err != nil {
+			return nil, err
+		}
+		return metricValue, nil
+	}
+	// This code is unreacheable
+	return nil, fmt.Errorf("Illegal state")
+}
+
+func (t *Translator) createFilter(objectIDs []string, project string, zone string, clusterName string, metricName string) (string, error) {
+	// project_id, cluster_name and pod_id together identify a pod unambiguously
+	projectFilter := fmt.Sprintf("resource.label.project_id = \"%s\"", project)
+	zoneFilter := fmt.Sprintf("resource.label.zone = \"%s\"", zone)
+	clusterFilter := fmt.Sprintf("resource.label.cluster_name = \"%s\"", strings.TrimSpace(clusterName))
+	// container_name is set to empty string for pod metrics
+	containerFilter := "resource.label.container_name = \"\""
+	metricFilter := fmt.Sprintf("metric.type = \"%s/%s\"", t.config.MetricsPrefix, metricName)
+
+	var nameFilter string
+	if len(objectIDs) == 0 {
+		return "", fmt.Errorf("No pods matched for metric %s", metricName)
+	} else if len(objectIDs) == 1 {
+		nameFilter = fmt.Sprintf("resource.label.pod_id = %s", objectIDs[0])
+	} else {
+		nameFilter = fmt.Sprintf("resource.label.pod_id = one_of(%s)", strings.Join(objectIDs, ","))
+	}
+	return fmt.Sprintf("(%s) AND (%s) AND (%s) AND (%s) AND (%s) AND (%s)", metricFilter, projectFilter, clusterFilter, zoneFilter, nameFilter, containerFilter), nil
+}
+
+func (t *Translator) getMetricValuesFromResponse(groupResource schema.GroupResource, namespace string, response *stackdriver.ListTimeSeriesResponse) (map[string]int64, error) {
+	group, err := api.Registry.Group(groupResource.Group)
```

Comment:
It merged the other day: https://github.com/kubernetes/kubernetes/pull/45431.  I just have yet to update the glide deps back to the official version.